manifest: update cmsis-nn to v6.0.0 & tflite-micro

JordanYates opened the pull request:
Update cmsis-nn to version 6.0.0.
Revision history can be found at:
https://arm-software.github.io/CMSIS-NN/latest/rev_hist.html

New upstream revision is the v6.0.0 release tag + module.yml https://github.com/ARM-software/CMSIS-NN/commit/69823017d0973440e311dcf900b2162b17609fb0

JordanYates committed:
manifest: update cmsis-nn to v6.0.0

Update cmsis-nn to version 6.0.0.
Revision history can be found at:
https://arm-software.github.io/CMSIS-NN/latest/rev_hist.html

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
tests: lib: cmsis_nn: add missing filter dimension

Specify the missing filter dimension that was the reason for the v6.0.0
version bump.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
doc: release-notes: document CMSIS-NN upgrade

Update the version update and link to the changelog.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented:
Updating CMSIS-NN also requires updating `tflite-micro`.
@XenuIsWatching do you remember how you generated the `third_party_static` folders last time?

XenuIsWatching commented:
> Updating CMSIS-NN also requires updating `tflite-micro`. @XenuIsWatching do you remember how you generated the `third_party_static` folders last time?

Yes, I put the info on how those were obtain in the commit message such as here https://github.com/zephyrproject-rtos/tflite-micro/commit/c3310510d9bacfdbaeac19307213d1631eb9b2c1

JordanYates commented:
> > Updating CMSIS-NN also requires updating `tflite-micro`. @XenuIsWatching do you remember how you generated the `third_party_static` folders last time?
> 
> Yes, I put the info on how those were obtain in the commit message such as here [zephyrproject-rtos/tflite-micro@c331051](https://github.com/zephyrproject-rtos/tflite-micro/commit/c3310510d9bacfdbaeac19307213d1631eb9b2c1)

```
Origin: kissfft
License: BSD-3-Clause
URL: https://github.com/mborgerding/kissfft/archive/refs/tags/v130.zip
Commit: 918336beac80868ad3011cb283063bb224e3e5da
Patch File: third_party/kissfft/kissfft.patch
Purpose: Used by TensorFlow Lite Micro
Maintained-by: External
```
Tells me what was imported, but not how. Did you build some sample application which pulled down all the sources which were then copied, or did you manually find git commit hashes and download sources from Github? Where did the patch files come from?

XenuIsWatching commented:
> > > Updating CMSIS-NN also requires updating `tflite-micro`. @XenuIsWatching do you remember how you generated the `third_party_static` folders last time?
> > 
> > 
> > Yes, I put the info on how those were obtain in the commit message such as here [zephyrproject-rtos/tflite-micro@c331051](https://github.com/zephyrproject-rtos/tflite-micro/commit/c3310510d9bacfdbaeac19307213d1631eb9b2c1)
> 
> ```
> Origin: kissfft
> License: BSD-3-Clause
> URL: https://github.com/mborgerding/kissfft/archive/refs/tags/v130.zip
> Commit: 918336beac80868ad3011cb283063bb224e3e5da
> Patch File: third_party/kissfft/kissfft.patch
> Purpose: Used by TensorFlow Lite Micro
> Maintained-by: External
> ```
> 
> Tells me what was imported, but not how. Did you build some sample application which pulled down all the sources which were then copied, or did you manually find git commit hashes and download sources from Github? Where did the patch files come from?

These are the same files that are pulled in as apart of the tflite-micro build process.
https://github.com/tensorflow/tflite-micro/tree/main/tensorflow/lite/micro/tools/make

For example you can find the kissfft within here which is downloaded, unzipped, and had the patch file applied
https://github.com/tensorflow/tflite-micro/blob/71fba59b5bebfd64b48e167bfcb669fff568ec74/tensorflow/lite/micro/tools/make/kissfft_download.sh#L48

Skimming through the folder... I noticed that if you update tflite-micro, it looks like flatbuffers will need to be updated along with the patch file applied
https://github.com/tensorflow/tflite-micro/pull/2274/files#diff-c7caf59f2356917e38d2acda5785d137d47d5ec33302d219795efe88440b3b49

They were all manually pulled down from the download links

XenuIsWatching commented:
As something I just found digging though... it looks like they pull a specific 'hash' of cmsis-nn which is this https://github.com/tensorflow/tflite-micro/blob/71fba59b5bebfd64b48e167bfcb669fff568ec74/tensorflow/lite/micro/tools/make/ext_libs/cmsis_download.sh#L50

It only looks like that hash is 2 commits behind the v6.0.0, so it 'may' still work with the v6.0.0

As I typed this... I noticed those 2 commits are only updates to pdsc files :/

JordanYates committed:
submanifests: update `tflite-micro`

Update `tflite-micro` to the latest version.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented:
Included the `tflite-micro` in this PR. I don't have permissions for the forked repo, and I don't think merging my commit into `main` or `zephyr` is the right approach. Would prefer a `zephyr-v3.7` branch or similar.

JordanYates committed:
modules: cmsis-nn: add `*_s4.c` sources

Add new sources for 4 bit function variants to build system.

Signed-off-by: Jordan Yates <jordan@embeint.com>

XenuIsWatching commented:
> Included the `tflite-micro` in this PR. I don't have permissions for the forked repo, and I don't think merging my commit into `main` or `zephyr` is the right approach. Would prefer a `zephyr-v3.7` branch or similar.
> 
> Actual branch this is pointing at: https://github.com/JordanYates/tflite-micro/commits/zephyr-v3.7/

looks like all the CI is passing now 👍 but I don't have admin rights to create a branch for `zephyr-3.7`. We may have to poke around to get that created

kartben commented:
with this PR, the hello world sample has a different behavior with regards to printf for floats

3.7 main:

```
SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..
*** Booting Zephyr OS build v3.7.0-rc1-266-g381ce06c126b ***
x_value: 1.0*2^-127, y_value: 1.0*2^-127

x_value: 1.2566366*2^-2, y_value: 1.4910772*2^-2

x_value: 1.2566366*2^-1, y_value: 1.1183078*2^-1

x_value: 1.8849551*2^-1, y_value: 1.677462*2^-1

x_value: 1.2566366*2^0, y_value: 1.9316229*2^-1
```

This PR:

```
SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..
*** Booting Zephyr OS build v3.6.0-6488-g8963be4230ec ***
x_value: *float*, y_value: *float*

x_value: *float*, y_value: *float*

x_value: *float*, y_value: *float*

x_value: *float*, y_value: *float*

x_value: *float*, y_value: *float*

x_value: *float*, y_value: *float*
```

Setting `CONFIG_REQUIRES_FLOAT_PRINTF=y` does the trick 

```
SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..
*** Booting Zephyr OS build v3.6.0-6488-g8963be4230ec ***
x_value: 0.000000, y_value: 0.000000

x_value: 0.314159, y_value: 0.372770

x_value: 0.628319, y_value: 0.559154

x_value: 0.942478, y_value: 0.838731

x_value: 1.256637, y_value: 0.965812

x_value: 1.570796, y_value: 1.042060

x_value: 1.884956, y_value: 0.957340

x_value: 2.199115, y_value: 0.821787

x_value: 2.513274, y_value: 0.533738

x_value: 2.827433, y_value: 0.237217
```

but note how number are formatted as regular floats, not scientific notation like the output before, so I am not sure what to think :)

kartben commented:
@JordanYates could you look into my comment above? It might be nothing and hopefully not a regression, but at a minimum the hello world sample and/or README would need updating, it looks like

JordanYates committed:
samples: tflite-micro: add `REQUIRES_FLOAT_PRINTF`

Add `CONFIG_REQUIRES_FLOAT_PRINTF=y` to tflite-micro samples as in the
past the module was using its own `printf` implementation, but now uses
our logging infrastructure.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented:
> @JordanYates could you look into my comment above? It might be nothing and hopefully not a regression, but at a minimum the hello world sample and/or README would need updating, it looks like

Looks like at some point tflite-micro switched from using their own logging functions to using ours, hence the change to `*float*`, and the differing output format once `CONFIG_REQUIRES_FLOAT_PRINTF` is enabled.

Added the define to both samples which don't currently have it.

stephanosio commented on west.yml:
```diff
@@ -124,7 +124,7 @@ manifest:
       revision: 6489e771e9c405f1763b52d64a3f17a1ec488ace
       path: modules/lib/cmsis-dsp
     - name: cmsis-nn
-      revision: 0c8669d81381ccf3b1a01d699f3b68b50134a99f
+      revision: ea987c1ca661be723de83bd159aed815d6cbd430
```
Based on what we do for other modules (e.g. TF-M), it would make more sense to merge the CMSIS-NN v6.0.0 tag into the existing `zephyr` branch than to create and reference a separate `zephyr-v6.0.0` branch.

Note that merge commits are allowed in the CMSIS-NN module repo.

JordanYates replied:
What makes more sense is up to the repo maintainer to decide. For the other modules I control I create branches per upstream release, as I see more value in having a clean diff (that is possible to review) from the upsteam release than a linear commit history (e.g. https://github.com/zephyrproject-rtos/loramac-node).

stephanosio replied:
> What makes more sense is up to the repo maintainer to decide

Sure, as long as the maintainer sticks with only one method.

> For the other modules I control I create branches per upstream release, as I see more value in having a clean diff (that is possible to review) from the upsteam release than a linear commit history

That makes sense; but, that does not seem to be how upstream syncs are meant to be handled for the particular module (especially, looking at the branch name) -- for example, what is the `zephyr` branch supposed to be after `zephyr-v6.0.0` branch is created?

If you, as the maintainer of the CMSIS-NN module, decide to do this, it would only make sense to rename the existing `zephyr` branch accordingly.

On a side note, maybe we need a place to properly document how every module handles upstream syncs so that we do not end up with a mess like `zephyr`, `zephyr-v2.6` and `v4.7.0-zephyr`, as seen in the `loramac-node` module repository.

stephanosio commented on submanifests/optional.yaml:
```diff
@@ -46,10 +48,10 @@ manifest:
       groups:
         - optional
     - name: tflite-micro
-      revision: 1a34dcab41e7e0e667db72d6a40999c1ec9c510c
+      revision: e1eecbee18aa8f7f0ad7e90f8b335564eb52c13f
```
same here

JordanYates replied:
The branching strategy for `tflite-micro` is the responsibility of @laurenmurphyx64.
Please provide guidance as to whether you would prefer a PR against main or whether you will push as a new branch.

laurenmurphyx64 replied:
I personally have no preference. I haven't been involved in the maintenance of tflite-micro in a long time (probably should remove myself from the collaborator role), and feel it would be better to leave the decision / execution to the folks actively involved in tflite-micro PRs.

JordanYates replied:
> same here

Alright, I have made a PR against the `zephyr` branch, someone with permissions on the repo can choose which to use.
https://github.com/zephyrproject-rtos/tflite-micro/pull/5
I wish the reviewers well with the 300k line, 1300 commit PR :)

ithinuel replied:
I started reviewing https://github.com/zephyrproject-rtos/tflite-micro/pull/5, I plan to publish that at some time tomorrow (once I finish double cross-checking my notes)

XenuIsWatching replied:
I just did an approval there, but I just have a 'shadow'ed' checkmark there...

I noticed that PR has 'merge conflicts' with that branch as the `zephyr` branch had it's own commits stacked on an older commit which was the latest at that time. I dunno if it's possible to merge it through Github....

Last time I updated tflite-micro, I had to ping around and @nashif did the merge of the branch there

JordanYates replied:
The whole merge conflicts issue will only get worse with time, hence my preference to go with a clean branch.
Anyway, the original PR was targeting the `zephyr` branch, which was wrong.
See https://github.com/zephyrproject-rtos/tflite-micro/pull/6 for the correct target.

ithinuel replied:
I left a change-request review there as there are a few things that are off although I don’t know how much it actually impacts the build.

stephanosio replied:
> The whole merge conflicts issue will only get worse with time, hence my preference to go with a clean branch.

Anyone opposed to this approach? I personally prefer starting with clean branches as well.

Now the question is, `zephyr-vZEPHYRVER` vs `vTFLITEMICROVER-zephyr` ...

ithinuel commented on submanifests/optional.yaml:
```diff
@@ -46,10 +48,10 @@ manifest:
+      revision: e1eecbee18aa8f7f0ad7e90f8b335564eb52c13f
       path: optional/modules/lib/tflite-micro
       repo-path: tflite-micro
-      remote: upstream
```
Is it ok to do this ?

stephanosio replied:
> Is it ok to do this ?

It is not. AFAICS, it is only supposed to be a temporary measure until we decide on how to handle upstream sync for the tflite-micro module.

ithinuel reviewed (dismissed):
As far as I understand this should still be pointing at the upstream (zephyr's fork) using the PR reference rather than a different remote.

nashif commented:
Can we please get issues resolved before RC2 on Friday, this PR shall go in for 3.7 as per the TSC meeting today.

nashif commented on submanifests/optional.yaml:
```diff
@@ -46,7 +46,7 @@ manifest:
         - optional
     - name: tflite-micro
-      revision: 1a34dcab41e7e0e667db72d6a40999c1ec9c510c
+      revision: /pull/6/head
```
need to remove leading slash /

nashif replied:
update to e1eecbee18aa8f7f0ad7e90f8b335564eb52c13f please

ithinuel replied:
based on the reviews and updates in https://github.com/zephyrproject-rtos/tflite-micro/pull/6,

the sha should be 48613f7ba1ffbda46ad771a77a35408f48f922e9 instead. (https://github.com/zephyrproject-rtos/tflite-micro’s `zephyr` branch should be updated accordingly)

JordanYates replied:
@ithinuel is correct, the original export commit had some duplicated files, please repush the last commit @nashif

nashif replied:
@JordanYates 

done
